Rewrite asset event registration

uranusjr opened the pull request:
There are many issues in the original implementation when you send something that's not a simple asset, I ended up almost rewritten the whole thing...

Now, the task runner collects everything the user writes into outlet_events, and send all of them with the task's outlets as declared by the user verbatim to the API server. The server does all the resolution and filtering instead.

Fix #46958 (in combination with https://github.com/apache/airflow/pull/47659).

uranusjr committed:
Rewrite asset event registration

There are many issues in the original implementation when you send
something that's not a simple asset, I ended up almost rewritten the
whole thing...

Now, the task runner collects everything the user writes into
outlet_events, and send all of them with the task's outlets as declared by the
user verbatim to the API server. The server does all the resolution and
filtering instead.

uranusjr committed:
Reuse code in execution_time

uranusjr committed:
Fix tests on both sides of the API

The task runner now emits slightly different data; extra is no longer
eagerly resolved; the server side fills in the default automatically.
Fixtures using asset refs now work correctly.

The API server side tests are vastly improved to cover more kinds of
data, including where events against other aliases are not incorrectly
picked up when processing events originated from an alias.

uranusjr committed:
Remove redundant arg from register_asset_change

The 'aliases' and 'source_alias_names' basically serve the same purpose,
but we use each for a different section of the code. They can become
one.

uranusjr committed:
Check for inactive assets

Exceptions are now raised if an event is raised against inactive assets.

uranusjr committed:
Correctly activate asset in test

uranusjr committed:
Ignore events against undeclared aliases

uranusjr committed:
Fix 2.x compat in Standard provider test

Lee-W commented on airflow/exceptions.py:
```diff
@@ -113,33 +113,42 @@ class AirflowFailException(AirflowException):
 
 
-class AirflowExecuteWithInactiveAssetExecption(AirflowFailException):
+class _AirflowExecuteWithInactiveAssetExecption(AirflowFailException):
```
Looks like we don't want anyone to use it directly. We could probably add this to the docstring.

uranusjr replied:
I mean in general we don’t want the user to catch most (any?) exceptions here anyway?

ashb replied:
And the underscore prefix is a pretty clear sign to not use it directly

Lee-W replied:
Yep, I just worried contributors might use it somewhere (most likely not, but...)

Lee-W commented on airflow/exceptions.py:
```diff
@@ -113,33 +113,42 @@ class AirflowFailException(AirflowException):
-        return ", ".join(
-            f'Asset(name="{key.name}", uri="{key.uri}")' for key in self.inactive_asset_unique_keys
-        )
+    def inactive_assets_message(self):
```
```suggestion
    def inactive_assets_message(self) -> str:
```

nit

Lee-W commented on airflow/models/taskinstance.py:
```diff
@@ -2744,104 +2733,126 @@ def _run_raw_task(
+                    session=session,
+                )
+
+        if bad_asset_keys:
```
if we check the condition here, does that mean we want handle good alias when bad asset key exists

uranusjr replied:
Not following

Lee-W replied:
If my understanding is correct, here is what we trying to do

1. collect assets and register the good ones
2. see if there are bad assets, if so, raise an exception
3. collect asset aliases and register the good ones
4. see if there are bad asset aliases; if so, raise an exception

so my question is "will it stop at step 2 while we probably should also run step 3"

uranusjr replied:
In practice 2 can’t fire unless you send a faulty request. I guess I can move it to the last instead? (Or before 4, but it’s easier to move it to the last since 3 and 4 are in the same if block.)

Lee-W replied:
yep, maybe let's move it to the last. 👍 thanks!

Lee-W commented on task-sdk/tests/task_sdk/execution_time/test_task_runner.py:
```diff
@@ -864,18 +832,7 @@ def test_run_with_asset_inlets(create_runtime_ti, mock_supervisor_comms):
-                        "key": {"name": "new-name", "uri": "s3://bucket/my-task"},
-                    },
-                ],
+                outlet_events=[],
```
Why do we remove the events?

uranusjr replied:
`outlet_events` now only include entries that don’t have the default extra. None of these have custom extras so they are not included in this list anymore. There is a test making sure outlets without an entry in `outlet_events` are still emitted correctly.

Lee-W replied:
got it. thanks!

ashb commented on task-sdk/src/airflow/sdk/definitions/asset/__init__.py:
```diff
@@ -60,7 +60,7 @@
 
 @attrs.define(frozen=True)
-class AssetUniqueKey:
+class AssetUniqueKey(attrs.AttrsInstance):
```
What this base class for?

uranusjr replied:
To convince Mypy we can do `attrs.asdict` to an instance. It does nothing at runtime.

uranusjr committed:
Trigger alias events when invalid assets found

This does not really matter in practice since inactive assets should
have been caught before the task runs, but in case any are found, we can
still try to trigger as many valid events as possible before failing the
task.

uranusjr committed:
Nits on exception classes
